test/common/buffer/owned_impl_test.cc

```diff
@@ -1263,6 +1265,218 @@ TEST_F(OwnedImplTest, FrontSlice) {
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+
+  buffer_two.drain(buffer_two.length());
+  EXPECT_EQ(buffer_one_account->balance(), 0);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+}
+
+TEST(BufferMemoryAccountTest,
+     SliceRemainsAttachToOriginalAccountWhenMovedUnlessCoalescedIntoExistingSlice) {
+  std::shared_ptr<BufferMemoryAccount> buffer_one_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_one;
+  buffer_one.bindAccount(buffer_one_account);
+  ASSERT_EQ(buffer_one_account->balance(), 0);
+
+  std::shared_ptr<BufferMemoryAccount> buffer_two_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_two;
+  buffer_two.bindAccount(buffer_two_account);
+  ASSERT_EQ(buffer_two_account->balance(), 0);
+
+  buffer_one.add("Will Coalesce");
+  buffer_two.add("To be Coalesce into:");
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 4096);
+
+  // Transfer over buffer, slices coalesce, crediting account one.
+  buffer_two.move(buffer_one);
+  EXPECT_EQ(buffer_one_account->balance(), 0);
+  EXPECT_EQ(buffer_two_account->balance(), 4096);
+
+  buffer_two.drain(std::string("To be Coalesce into:Will Coalesce").length());
+  EXPECT_EQ(buffer_one_account->balance(), 0);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+}
+
+TEST(BufferMemoryAccountTest, SliceCanRemainAttachedToOriginalAccountWhenMovedAndCoalescedInto) {
+  std::shared_ptr<BufferMemoryAccount> buffer_one_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_one;
+  buffer_one.bindAccount(buffer_one_account);
+  ASSERT_EQ(buffer_one_account->balance(), 0);
+
+  std::shared_ptr<BufferMemoryAccount> buffer_two_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_two;
+  buffer_two.bindAccount(buffer_two_account);
+  ASSERT_EQ(buffer_two_account->balance(), 0);
+
+  std::shared_ptr<BufferMemoryAccount> buffer_three_account =
+      std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_three;
+  buffer_three.bindAccount(buffer_three_account);
+  ASSERT_EQ(buffer_three_account->balance(), 0);
+
+  buffer_one.add("Will Coalesce");
+  buffer_two.add("To be Coalesce into:");
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 4096);
+
+  // Transfer buffers, leading to slice coalescing in third buffer.
+  buffer_three.move(buffer_two);
+  buffer_three.move(buffer_one);
+  EXPECT_EQ(buffer_one_account->balance(), 0);
+  EXPECT_EQ(buffer_two_account->balance(), 4096);
+  EXPECT_EQ(buffer_three_account->balance(), 0);
+
+  buffer_three.drain(std::string("To be Coalesce into:Will Coalesce").length());
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+}
+
+TEST(BufferMemoryAccountTest, LinearizedBufferShouldChargeItsAssociatedAccount) {
+  std::shared_ptr<BufferMemoryAccount> buffer_one_account = std::make_shared<BufferMemoryAccount>();
```

Comment:
nit: it is often preferable to use auto as the variable type when creating an object using std::make_shared

auto buffer_one_account = std::make_shared<BufferMemoryAccount>();